### ISSUE TITLE
[RFC] dune-package files

### ISSUE THREAD

**Issue by diml** [ocaml/dune#1206]
The metadata for installed libraries are starting to be a bit scattered (META files, `<lib>.dune files`, ...). I suggest  to factorize them in a single `dune-package` file. This file would use the same versioning mechanism as `dune-project` files and would contain all the metadata dune needs to know about.

## Overview

When looking for a external library `foo` or `foo.*`, dune would look for a file `foo/dune-package` in the library search path. If no such file can be found, dune would look for a plain findlib package instead. Effectively, when an external directory would contain both a `dune-package` and a `META` file, the `META` file would be ignored by dune.

Dune would generate one `dune-package` file for every package it knows about and would install it in the library directory of the package. It would continue to generate and install one `META` file as well, for compatibility with non dune build systems.

## Example of contents

```
(lang dune 1.0)
(using in_development_do_not_use_variants 0.0)

(library
 (name foo.bar)
 (dir bar)
 (libraries a b c)
 (modules ...)
 (virtual_modules ...))

(library
 (name foo.baz)
 (dir baz)
 (inline_tests.backend ...))
```

The `library` stanza in this file would basically be a direct representation of the `Lib.Info.t` type.

**Comment by rgrinberg** [comment]
@ejgallego this ticket might be of interest to you. If dune were to gain support for building coq projects, this is the file where you'd include w/e metadata you needed for coq libraries.

**Comment by rgrinberg** [comment]
Currently, dune searches for external libraries in one of these places:

* `$ ocamlfind printconf path`
* `$ opam config var lib`
* hard coded path relative to `ocamlc`

These choices were made to cover the common installation scenarios for users. There's no reason why we can't extend dune to look for libraries wherever coq users place them.

**Comment by andreypopp** [comment]
> Currently, dune searches for external libraries in one of these places:

This is one of the reasons we have a custom override for dune to work with esy — we inject `ocamlfind` as a dependency to `dune` so that it does `ocamlfind printconf path` (it can't do `opam config var lib` for obvious reasons and hard coded paths relative to `ocamlc` won't work as `esy` installs packages into separate prefixes) for which we craft `OCAMLPATH` env var.

It would be really great if `dune` supported `OCAMLPATH` directly.

**Comment by rgrinberg** [comment]
@andreypopp would it use OCAMLPATH when `ocamlfind` and `opam` were absent?

**Comment by andreypopp** [comment]
Yes, something like that:
- `$ ocamlfind printconf path`
- `$ opam config var lib`
- `$OCAMLPATH`
- hard coded path relative to ocamlc

**Comment by diml** [comment]
The difficulty is to come up with a scheme that works for most environments. I suppose we can add a dune configure time option to select how dune discovers the search path.

For instance `./configure --ocaml-library-path-method [OCAMLPATH|opam|...]`. When it's `OCAMLPATH` it would only look at the `OCAMLPATH` environment variable.

BTW, we could stop calling `ocamlfind` or `opam config var lib` when `OPAM_SWITCH_PREFIX` is set.

**Comment by rgrinberg** [comment]
Why do we need a separate configure phase however? Seems like we could control this behavior quite easily via:

* Env var
* command line option
* Context configuration in workspace

**Comment by diml** [comment]
> For instance ./configure --ocaml-library-path-method [OCAMLPATH|opam|...]. When it's OCAMLPATH it would only look at the OCAMLPATH environment variable.

Or this could also be done via an environment variable. Having less configure time options allows to make the produced binary more generic.

**Comment by diml** [comment]
@rgrinberg indeed, I just came to the same conclusion.

**Comment by diml** [comment]
I wrote #1226 to clean up a bit the detection of the library search path. In particular, the only case where we still do `opam config var lib` is when using `(context (opam ...))` in the dure-workspace file and using opam1 instead of opam2.

**Comment by andreypopp** [comment]
Thank you! This makes it much cleaner for us.

**Comment by rgrinberg** [comment]
@diml the `dir` field that you have in your example - is it really necessary? I think it would be better if the directory structure just followed the library names. Also, there's no dir field in the info type so it's a bit awkward to have it as well.

**Comment by rgrinberg** [comment]
Fixed in 1.7
